examples/step-57/step-57.cc

```diff
@@ -0,0 +1,948 @@
+                                                 + grad_phi_u[j]*present_velocity_values[q]*phi_u[i] //(U_old*grad_d_U, V_u)
+                                                 - div_phi_u[i]*phi_p[j]                             //-(d_P, div_V_u)
+                                                 - phi_p[i]*div_phi_u[j]                             //-(div_d_U, V_p)
+                                                 + gamma*div_phi_u[j]*div_phi_u[i]                   //grad-div (div_u_j, div_u_i)
+                                                 + phi_p[i]*phi_p[j] )                               //Mp-mass matrix of pressure
+                                              * fe_values.JxW(q);
+
+                      }
+                  }
+
+                if (assemble_rhs)
+                  {
+                    local_rhs(i) += (
+                                      -viscosity*scalar_product(present_velocity_gradients[q],grad_phi_u[i]) // -(gradU_old, gradV_u)
+                                      -present_velocity_gradients[q]*present_velocity_values[q]*phi_u[i]
+                                      // -(U_old*gradU_old, V_u)
+                                      +present_pressure_values[q]*div_phi_u[i] // +(P_old, div_V_u)
+                                      +present_velocity_divergence[q]*phi_p[i]   // +(div_U_oid, V_p)
+                                      -gamma*present_velocity_divergence[q]*div_phi_u[i]
+                                    )*fe_values.JxW(q);
+                  }
+
+              }
+          }
+
+        cell-> get_dof_indices (local_dof_indices);
+
+        if (initial_step)
+          {
+            if (assemble_matrix)
+              {
+                nonzero_constraints.distribute_local_to_global(local_matrix,
+                                                               local_dof_indices,
+                                                               system_matrix);
+              }
+
+            if (assemble_rhs)
+              {
+                nonzero_constraints.distribute_local_to_global(local_rhs,
+                                                               local_dof_indices,
+                                                               system_rhs);
+              }
+          }
+
+        else
+          {
+            if (assemble_matrix)
+              {
+                zero_constraints.distribute_local_to_global(local_matrix,
+                                                            local_dof_indices,
+                                                            system_matrix);
+              }
+
+            if (assemble_rhs)
+              {
+                zero_constraints.distribute_local_to_global(local_rhs,
+                                                            local_dof_indices,
+                                                            system_rhs);
+              }
```

Comment:
Can't you define a reference `const ConstraintMatrix &constraints_used = initial_step ? nonzero_constraints : zero_constraints`. This way you avoid duplicating all the `distribute_local_to_global` calls.